src/util/global-config.js

```diff
@@ -0,0 +1,120 @@
+const debug = require('./debug').here(__filename);
+const { resolve } = require('path');
+const { homedir } = require('os');
+const { createHash } = require('crypto');
+const flatfile = require('flat-file-db');
+
+class GlobalConfig {
+    static getDbFilePath() {
+        return resolve(homedir(), './.config/pwa-buildpack.db');
+    }
+    static async db() {
+        if (!this._dbPromise) {
+            this._dbPromise = new Promise((joy, pain) => {
+                try {
+                    const dbFilePath = this.getDbFilePath();
+                    debug(`no cached db exists, pulling db from ${dbFilePath}`);
+                    const db = flatfile(dbFilePath);
+                    debug(`db created, waiting for open event`, db);
+                    db.on('open', () => {
+                        debug('db open, fulfilling to subscribers');
+                        joy(db);
+                    });
+                } catch (e) {
+                    pain(e);
+                }
+            });
+        }
+        return this._dbPromise;
+    }
+    constructor(options) {
+        // validation
+        if (typeof options !== 'object') {
+            throw Error(debug.errorMsg('Must provide options.'));
+        }
+
+        const { key, prefix } = options;
+
+        if (typeof key !== 'function') {
+            throw Error(
+                debug.errorMsg('Must provide a `key` function in options.')
+            );
+        }
+
+        if (typeof prefix !== 'string') {
+            throw Error(
+                debug.errorMsg('Must provide a `prefix` string in options.')
+            );
+        }
+
+        this._makeKey = key;
+        this._prefix = prefix;
+    }
+    makeKey(keyparts) {
+        if (keyparts.length !== this._makeKey.length) {
+            throw Error(
+                `Wrong number of arguments sent to produce unique ${
+                    this._prefix
+                } key`
+            );
+        }
+        if (keyparts.length === 0) {
+            // a scalar, then
```

Comment:
I know what a scalar is, but I'm not sure what this comment is conveying to me

Reply:
It was a bad and confusing corner case for the GlobalConfig object, which acted like a single-value store if the key function had no parameters. I've removed it!